rust_src/src/editfns.rs

```diff
@@ -1447,4 +1450,284 @@ pub fn widen() {
+/// (see `unibyte-char-to-multibyte').  If the current buffer is unibyte, multibyte strings are
+/// converted to unibyte for insertion.
+///
+/// usage: (insert-and-inherit &rest ARGS)
+#[lisp_fn(name = "insert-and-inherit", c_name = "insert_and_inherit")]
+pub fn insert_and_inherit_lisp(args: &[LispObject]) {
+    general_insert_function(insert_and_inherit, insert_from_string_safe, true, args);
+}
+
+/// Insert strings or characters at point, relocating markers after the text.  Point and markers
+/// move forward to end up after the inserted text.
+///
+/// If the current buffer is multibyte, unibyte strings are converted to multibyte for insertion
+/// (see `unibyte-char-to-multibyte').  If the current buffer is unibyte, multibyte strings are
+/// converted to unibyte for insertion.
+///
+/// If an overlay begins at the insertion point, the inserted text falls outside the overlay; if a
+/// nonempty overlay ends at the insertion point, the inserted text falls inside that overlay.
+///
+/// usage: (insert-before-markers &rest ARGS)
+#[lisp_fn(name = "insert-before-markers", c_name = "insert_before_markers")]
+pub fn insert_before_markers_lisp(args: &[LispObject]) {
+    general_insert_function(
+        insert_before_markers,
+        insert_from_string_before_markers_safe,
+        false,
+        args,
+    );
+}
+
+/// Insert text at point, relocating markers and inheriting properties.  Point and markers move
+/// forward to end up after the inserted text.
+///
+/// If the current buffer is multibyte, unibyte strings are converted to multibyte for insertion
+/// (see `unibyte-char-to-multibyte').  If the current buffer is unibyte, multibyte strings are
+/// converted to unibyte for insertion.
+///
+/// usage: (insert-before-markers-and-inherit &rest ARGS)
+#[lisp_fn(
+    name = "insert-before-markers-and-inherit",
+    c_name = "insert_and_inherit_before_markers"
+)]
+pub fn insert_and_inherit_before_markers_lisp(args: &[LispObject]) {
+    general_insert_function(
+        insert_before_markers_and_inherit,
+        insert_from_string_before_markers_safe,
+        true,
+        args,
+    );
+}
+
+/// Insert the part of the text of STRING, a Lisp object assumed to be of type string, consisting
+/// of the LENGTH characters (LENGTH_BYTE bytes) starting at position POS / POS_BYTE.  If the text
+/// of STRING has properties,
+///copy them into the buffer.
+///
+/// It does not work to use `insert' for this, because a GC could happen before we copy the stuff
+/// into the buffer, and relocate the string without insert noticing.
+#[no_mangle]
+pub unsafe extern "C" fn insert_from_string(
+    string: LispObject,
+    pos: ptrdiff_t,
+    pos_byte: ptrdiff_t,
+    length: ptrdiff_t,
+    length_byte: ptrdiff_t,
+    inherit: bool,
+) {
+    let current_buffer = ThreadState::current_buffer_unchecked();
+    let opoint = current_buffer.pt;
+
+    let s = string.as_string_or_error();
```

Comment:
this can be `let s: LispStringRef = string.into();`. See @nickdrozd's recent PRs.

Reply:
Changed and fixed up.